cmd/update-main.go

```diff
@@ -31,204 +33,254 @@ import (
+		return releaseTime, probe.NewError(e)
 	}
 
-	if current.IsZero() {
-		err = errDummy().Trace(newUpdateURL)
-		errMsg = "Updates not supported for custom builds. Version field is empty. Please download official releases from https://minio.io/#minio"
-		return updateMessage{}, errMsg, err
+	if !(fields[0] == "mc" && fields[1] == "RELEASE") {
+		e := fmt.Errorf("Unknown release '%s'", releaseInfo)
+		return releaseTime, probe.NewError(e)
 	}
 
-	body, e := ioutil.ReadAll(data.Body)
+	var e error
+	releaseTime, e = time.Parse(releaseTagTimeLayout, fields[2])
 	if e != nil {
-		err = probe.NewError(e)
-		errMsg = "Fetching updates failed. Please try again."
-		return updateMessage{}, errMsg, err
+		e = fmt.Errorf("Unknown release time format. %s", err)
 	}
 
-	latest, err := parseReleaseData(string(body))
+	return releaseTime, probe.NewError(e)
+}
+
+func getLatestReleaseTime(timeout time.Duration) (releaseTime time.Time, err *probe.Error) {
+	data, err := DownloadReleaseData(timeout)
 	if err != nil {
-		errMsg = "Please report this issue at https://github.com/minio/mc/issues."
-		return updateMessage{}, errMsg, err.Trace(newUpdateURL)
+		return releaseTime, err.Trace()
 	}
+	return parseReleaseData(data)
+}
+
+func getDownloadURL() (downloadURL string) {
+	if IsDocker() {
+		return "docker pull minio/mc"
+	}
+
+	if runtime.GOOS == "windows" {
+		return mcReleaseURL + "mc.exe"
+	}
+
+	return mcReleaseURL + "mc"
+}
 
-	if latest.IsZero() {
-		err = errDummy().Trace(newUpdateURL)
-		errMsg = "Unable to validate any update available at this time. Please open an issue at https://github.com/minio/mc/issues"
-		return updateMessage{}, errMsg, err
+func getUpdateInfo(timeout time.Duration) (older time.Duration, downloadURL string, err *probe.Error) {
+	currentReleaseTime, err := GetCurrentReleaseTime()
+	if err != nil {
+		return older, downloadURL, err.Trace()
 	}
 
-	updateMsg = updateMessage{
-		Download: downloadURL,
-		Version:  Version,
+	latestReleaseTime, err := getLatestReleaseTime(timeout)
+	if err != nil {
+		return older, downloadURL, err.Trace()
 	}
-	if latest.After(current) {
-		updateMsg.Update = true
+
+	if latestReleaseTime.After(currentReleaseTime) {
+		older = latestReleaseTime.Sub(currentReleaseTime)
+		downloadURL = getDownloadURL()
 	}
-	return updateMsg, "", nil
+
+	return older, downloadURL, nil
 }
 
-// main entry point for update command.
-func mainUpdate(ctx *cli.Context) error {
+func mainUpdate(ctx *cli.Context) {
+	if len(ctx.Args()) != 0 {
+		cli.ShowCommandHelpAndExit(ctx, "update", -1)
+	}
 
 	// Additional command speific theme customization.
 	console.SetColor("Update", color.New(color.FgGreen, color.Bold))
 
-	var updateMsg updateMessage
-	var errMsg string
-	var err *probe.Error
-	// Check for update.
-	if ctx.Bool("experimental") {
-		updateMsg, errMsg, err = getReleaseUpdate(mcUpdateExperimentalURL)
-	} else {
-		updateMsg, errMsg, err = getReleaseUpdate(mcUpdateStableURL)
+	setGlobalsFromContext(ctx)
```

Comment:
We changed recently to setting setGlobalsFromContext in Command.Before, that way we can support global flags being inserted in any place.